### ISSUE TITLE
x/tools/go/packages: high CPU usage while processing cgo file

### ISSUE THREAD

**Issue by RadhiFadlillah** [golang/go#32821]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.12.5 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GOARCH="amd64"
GOBIN="/home/radhi/Go/bin"
GOCACHE="/home/radhi/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/radhi/Go"
GOPROXY=""
GORACE=""
GOROOT="/usr/lib/go"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build763115998=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

This issue is related with [vscode-go#2598](https://github.com/microsoft/vscode-go/issues/2598). While working on cgo project, both `gocode` and `gopls` will use huge amount of CPU when making autocomplete suggestion. It can be seen [here](https://giant.gfycat.com/WateryUnnaturalCrane.webm) for `gocode` and [here](https://giant.gfycat.com/DiligentUglyGrouper.webm) for `gopls`.

After looking around, I found out the reason for this is because both [`gocode`](https://github.com/stamblerre/gocode/blob/81059208699789f992bb4a4a3fedd734e335468d/internal/suggest/suggest.go#L138) and [`gopls`](https://github.com/golang/tools/blob/252024b8295926254bbc18903b9b7d4f0389df2f/internal/lsp/cache/view.go#L118) uses `LoadSyntax` mode, which make `go/packages` run `go list` with `-compiled` flag set to true, which lead to high CPU usage in cgo project.

### What did you expect to see?

The cgo file can be parsed as light as the normal Go code.

### What did you see instead?

The process uses huge amount of CPU.

**Comment by bcmills** [comment]
CC @matloob @ianthehat

**Comment by bcmills** [comment]
I'm going to assume for now that this is something that can be addressed within `go/packages`, but please let me and @jayconrod know if this is looking like a bug in `go list -compiled` instead.

**Comment by matloob** [comment]
When requesting source in go/packages, we always return the cgo-processed source code. That's where we chose to draw the abstraction boundary. So I don't think there's too much we can do from the go/packages side. We're going to have to run go list -compiled, and once go list runs, go/packages does minimal work to copy the outputs of go/packages into the Packages struct it returns in a Load call (since we don't need to do parsing/typechecking/etc).

The one strange thing is that it's my understanding that the LSP gets the package information only once (unless new packages are opened) so there shouldn't be  consistent high cpu usage.

**Comment by stamblerre** [comment]
`gopls` will run `go/packages` whenever a imports are changed, so it could be happening more frequently.

**Comment by segevfiner** [comment]
Probably related if not duplicate of https://github.com/golang/go/issues/29427

**Comment by pjweinb** [comment]
This is obsolete since gopls introduced Snaphsots and stopped running go commands during completions.
